### PR TITLE
Fixes to PR 364 with failing tests + assorted issues

### PR DESCRIPTION
**Pull request by EmilyBourne**
- Provide the allocated size to function arguments. Keep the same size if the size is a PyccelArraySize
- An unknown shape is a PyccelArraySize or None. Add this logic to PyccelBooleanOperator too
- Add tests that would have been failing before corrections
- Add error test to ensure that error is raised when an array is created from an expression with unknown size (see issue #352)
- Add tests for issue #369
- If an array is created from an expression and the size is known then allocate the memory with an empty expression. Fixes issue #352
- Use the process_dtype function to find Array's dtype and precision. Fixes issue #367 
- Check for Boolean type for List. This mimmicks what is done in the Tuple class. Fixes issue #368
- Fix issue #369. When casting a boolean to a boolean change the precision. When casting something else to a boolean use a merge statement

### CODE REVIEW

**Comment by EmilyBourne** [comment]
Should we fix issue #335 before merging this?

File: pyccel/parser/semantic.py [hunk 1]
```diff
@@ -1381,14 +1382,18 @@ def _assign_lhs_variable(self, lhs, d_var, rhs, **settings):
...
                 if not know_lhs_shape:
                     msg = "Cannot infer shape of right-hand side for expression {} = {}".format(lhs, rhs)
                     errors.report(PYCCEL_RESTRICTION_TODO+'\n'+msg,
                         bounding_box=(self._current_fst_node.lineno, self._current_fst_node.col_offset),
                         severity='fatal', blocker=self.blocking)
-
+                elif lhs.rank>0 and isinstance(rhs, PyccelOperator):
+                    #TODO: Provide order once issue #335 is fixed
+                    #stmts.append(Assign(lhs, Empty(lhs.alloc_shape, dtype, lhs.order)))
+                    new_expressions.append(Assign(lhs, Empty(lhs.alloc_shape, dtype, 'C')))
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:1396]: Should we fix issue #335 before merging this PR?

Comment (yguclu) [reply pyccel/parser/semantic.py:1396]: I don't know how long it would take to fix that issue. This PR looks good as it stands

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1396]: I would rather leave it for another PR. I was just worried about potentially introducing a bug here

File: pyccel/parser/semantic.py [hunk 2]
```diff
@@ -1381,14 +1382,18 @@ def _assign_lhs_variable(self, lhs, d_var, rhs, **settings):
...
                         or isinstance(rhs, (Variable, EmptyLike, DottedVariable))
                 if not know_lhs_shape:
                     msg = "Cannot infer shape of right-hand side for expression {} = {}".format(lhs, rhs)
                     errors.report(PYCCEL_RESTRICTION_TODO+'\n'+msg,
                         bounding_box=(self._current_fst_node.lineno, self._current_fst_node.col_offset),
                         severity='fatal', blocker=self.blocking)
-
+                elif lhs.rank>0 and isinstance(rhs, PyccelOperator):
+                    #TODO: Provide order once issue #335 is fixed
+                    #stmts.append(Assign(lhs, Empty(lhs.alloc_shape, dtype, lhs.order)))
```
Comment (yguclu) [inline pyccel/parser/semantic.py:1395]: I know this is just a comment, but nevertheless... shouldn't `stmts` be changed to `new_expressions`?

Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: You're right. I had stmts as a local variable at one point. I must have missed this when I changed it to new_expressions

File: tests/epyccel/modules/arrays.py [hunk 3]
```diff
@@ -364,6 +364,71 @@ def array_real_2d_F_initialization(a):
...
+    from numpy import full
+    z = full((2,3),5,order='F')
+    x[:] = (x // y) * x + z
+
+@types( 'int32[:]', 'int32[:]', 'int32[:]' )
+def array_int32_in_bool_out_1d_complex_3d_expr( x, y, ri ):
+    from numpy import full, int32, empty
+    z = full(3,5, dtype=int32)
+    r = empty(3, dtype=bool)
+    r = (x // y) * x > z
```
Comment (yguclu) [inline tests/epyccel/modules/arrays.py:413]: Is this the expression you intended, or maybe you meant to write `r[:] = ...`?

Comment (EmilyBourne) [reply tests/epyccel/modules/arrays.py]: You are right. It should be the same as in the functions below it

**Pull request merged**